nixos/modules/installer/tools/nixos-option/nixos-option.cc

```diff
@@ -0,0 +1,599 @@
+    for (auto i = tokens.begin(); i != tokens.end(); i++) {
+        bool last_attribute = std::next(i) == tokens.end();
+        auto const & attr = *i;
+        v = evaluateValue(ctx, &v);
+        if (attr.empty()) {
+            throw Error("empty attribute name in selection path '%s'", path);
+        }
+        if (isOption(ctx, v) && optionTypeIs(ctx, v, "submodule")) {
+            v = getSubOptions(ctx, v);
+        }
+        if (isOption(ctx, v) && optionTypeIs(ctx, v, "loaOf") && !last_attribute) {
+            v = getSubOptions(ctx, v);
+            // Note that we've consumed attr, but didn't actually use it.  This is the path component that's looked up
+            // in the list or attribute set that doesn't name an option -- the "root" in "users.users.root.name".
+        } else if (v.type != tAttrs) {
+            throw Error("attribute '%s' in path '%s' attempts to index a value that should be a set but is %s", attr,
+                        path, showType(v));
+        } else {
+            auto const & next = v.attrs->find(ctx->state->symbols.create(attr));
+            if (next == v.attrs->end()) {
+                throw Error("attribute '%s' in path '%s' not found", attr, path);
+            }
+            v = *next->value;
+        }
+    }
+    return v;
+}
+
+void printOne(Context * ctx, Out & out, std::string const & path)
+{
+    try {
+        Value option = findAlongOptionPath(ctx, path);
+        option = evaluateValue(ctx, &option);
+        if (isOption(ctx, option)) {
+            printOption(ctx, out, path, &option);
+        } else {
+            printListing(out, &option);
+        }
+    } catch (Error & e) {
+        std::cerr << "error: " << e.msg()
+                  << "\nAn error occurred while looking for attribute names. Are "
+                     "you sure that '"
+                  << path << "' exists?\n";
+    }
+}
+
+int main(int argc, char ** argv)
+{
+    bool all = false;
+    std::string path = ".";
+    std::string options_expr = "(import <nixpkgs/nixos> {}).options";
+    std::string config_expr = "(import <nixpkgs/nixos> {}).config";
+    std::vector<std::string> args;
+
+    struct MyArgs : nix::LegacyArgs, nix::MixEvalArgs
+    {
+        using nix::LegacyArgs::LegacyArgs;
+    };
+
+    MyArgs myArgs(nix::baseNameOf(argv[0]), [&](Strings::iterator & arg, const Strings::iterator & end) {
+        if (*arg == "--help") {
+            nix::showManPage("nixos-option");
+        } else if (*arg == "--version") {
+            nix::printVersion("nixos-option");
+        } else if (*arg == "--all") {
+            all = true;
+        } else if (*arg == "--path") {
+            path = nix::getArg(*arg, arg, end);
+        } else if (*arg == "--options_expr") {
+            options_expr = nix::getArg(*arg, arg, end);
+        } else if (*arg == "--config_expr") {
+            config_expr = nix::getArg(*arg, arg, end);
+        } else if (!arg->empty() && arg->at(0) == '-') {
```

Comment:
Maybe unrecognised options should be passed onto nix itself, e.g. to enable pure evaluation, set tarball ttl (when having URLs on the NIX_PATH for instance), or set `system` (to evaluate foreign-arch systems).

Reply:
I think this happens already by virtue of using `nix::LegacyArgs` and/or `nix::MixEvalArgs` above.  I observe that I can pass `-I nixpkgs=/some/other/path` to make it read the package definitions from elsewhere, even though there is no explicit handling of `-I` here.  (In fact, this is super-useful when used as `diff -u <(nixos-option --all) <(nixos-option --all -I nixpkgs=/some/other/path)`)